### PR TITLE
Report specific type of error when credentials can't be fetched

### PR DESCRIPTION
**Pull request by jurre**
This way, we can inform users what went wrong.

I've decided to try to stick to the error format we use in our internal
infrastructure as much as possible, without updating the JobError.
Because if that, this is accompanied by a small change to our API that
handles these errors.

### CODE REVIEW

File: src/api-client.ts [hunk 1]
```diff
@@ -56,7 +58,11 @@ export class ApiClient {
       headers: {Authorization: this.params.credentialsToken}
     })
     if (res.status !== 200) {
-      throw new Error(`Unexpected status code: ${res.status}`)
+      throw new CredentialFetchingError(
+        `fetching credentials: received code ${res.status}: ${JSON.stringify(
+          res.data
+        )}`
```
Comment (brrygrdn) [inline src/api-client.ts:64]: I wonder if we should only raise this error for `422` range errors so we can treat it as a client side, i.e. action configuration, problem?

It's not a huge deal right now, but in future we'll want to be specific about 5xx vs 4xx when it comes to instrumentation.

That said, we currently do not have any `DependabotErrorType` value defined for a configuration/input error so we can't do better without shipping service-side changes 🤔

Comment (jurre) [reply src/api-client.ts:64]: Definitely an option, I just mirrored what we do in our AWS based infrastructure. We look at the errors returned from the API to determine if this was an error with referencing the secrets so functionally I do think it would matter

Comment (mctofu) [reply src/api-client.ts:64]: I had a similar question from https://github.com/github/dependabot-api/pull/2615 about how we would handle non-422 errors here. The dependabot-update-job-runner does have some special handling of 4xx vs 5xx errors to decide whether we need to retry: https://github.com/github/dependabot-update-job-runner/blob/main/runner/sqsqueue/limit.go#L173. That part may not be super relevant to how we work on actions though.

One other thing is we'd only report the secret not found error to the user if the 422 error had the secret not found message in it. Other 4xx or 5xx errors would end up as unknown errors.

Comment (jurre) [reply src/api-client.ts:64]: @mctofu could you rephrase what you're suggesting? I don't think I follow entirely.

For context, this does not tackle any other errors than those for secrets not found, but most other errors were [already handled](https://github.com/dependabot/updater-action/blob/935d3219992ad2d3a103a719ff346b530e611e26/src/main.ts#L90) and don't require any special data to be added to be handled by dependabot-api as far as I could tell, and rather than try to hunt down every edge-case I focussed on this specific error.

Happy to tackle other potential issues either in this PR or separately though!

Comment (mctofu) [reply src/api-client.ts:64]: @jurre: After reviewing the API change I got the full picture and I think this looks good!

I would like to better understand how job retries are handled on actions and whether this job would be retried by us or actions on a 422 error from the credentials api. That would be best addressed separately from this change though.

**Review by mctofu — APPROVED** [review]

File: src/api-client.ts [hunk 2]
```diff
@@ -52,24 +54,34 @@ export class ApiClient {
 
   async getCredentials(): Promise<Credential[]> {
     const credentialsURL = `/update_jobs/${this.params.jobId}/credentials`
-    const res: any = await this.client.get(credentialsURL, {
-      headers: {Authorization: this.params.credentialsToken}
-    })
-    if (res.status !== 200) {
-      throw new Error(`Unexpected status code: ${res.status}`)
-    }
```
Comment (jurre) [inline src/api-client.ts:60]: Turns out that this didn't actually handle any errors, we should fix that for the rest of the api methods as well, but the result so far is mostly the same as axios throws an error very similar to the one that we're intending to throw

**Comment by jurre** [comment]
Ended up needing to make a few changes to the error handling code, so would appreciate another look at this

**Review by mctofu — APPROVED** [review]
